### PR TITLE
Added CLI to Instant Logs

### PR DESCRIPTION
**Pull request by angelampcosta**

### CODE REVIEW

File: products/logs/src/content/instant-logs.md [hunk 1]
```diff
@@ -5,9 +5,9 @@ pcx-content-type: concept
 
 # Instant Logs
 
-Instant Logs allows Cloudflare Enterprise customers to access a live stream of the traffic for their domain on the Cloudflare dashboard. Seeing data in real time allows you to investigate an attack, troubleshoot, debug or test out changes made to your network. Instant Logs is lightweight, simple to use and does not require any additional setup.
+Instant Logs allows Cloudflare Enterprise customers to access a live stream of the traffic for their domain from the Cloudflare dashboard or from a command-line interface (CLI). Seeing data in real time allows you to investigate an attack, troubleshoot, debug or test out changes made to your network. Instant Logs is lightweight, simple to use and does not require any additional setup.
 
-## How to use Instant Logs
+## Instant Logs on the Cloudflare Dashboard
```
Comment (marciocloudflare) [inline products/logs/src/content/instant-logs.md]:
```suggestion
## Instant Logs on the Cloudflare dashboard
```

File: products/logs/src/content/instant-logs.md [hunk 2]
```diff
@@ -19,10 +19,119 @@ Instant Logs allows Cloudflare Enterprise customers to access a live stream of t
...
+
+### 1. Create an Instant Logs Job
+
+Create a session by sending a `POST` request to our Instant Logs job endpoint with the following parameters:
+
+* **Fields** - List any field available in our [HTTP request dataset](https://developers.cloudflare.com/logs/reference/log-fields/zone/http_requests).
+
+* **Sample** - The sample parameter is the sample rate of the records set by the client: `"sample": 1` is 100% of records `"sample": 10` is 10% and so on.
+
+Note: Instant Logs has a maximum data rate supported. For high volume domains, we sample server side as indicated in the `"sampleInterval"` parameter returned in the logs.
```
Comment (marciocloudflare) [inline products/logs/src/content/instant-logs.md]:
```suggestion
<Aside type='note'>
Instant Logs has a maximum data rate supported. For high volume domains, we sample server side as indicated in the `"sampleInterval"` parameter returned in the logs.
</Aside>
```

File: products/logs/src/content/instant-logs.md [hunk 3]
```diff
@@ -19,10 +19,119 @@ Instant Logs allows Cloudflare Enterprise customers to access a live stream of t
...
+
+* **Filters** - Use filters to drill down into specific events. Filters consist of 3 parts: key, operator and value. The keys we support are **ASN**, **Cache status**, **Country**, **Client IP**, **Host**, **HTTP method**, **Path**, **Status code**, **Firewall action matches**, and **Firewall rule ID matches**.
```
Comment (marciocloudflare) [inline products/logs/src/content/instant-logs.md]:
```suggestion
* **Filters** - Use filters to drill down into specific events. Filters consist of three parts: key, operator and value. The keys we support are **ASN**, **Cache status**, **Country**, **Client IP**, **Host**, **HTTP method**, **Path**, **Status code**, **Firewall action matches**, and **Firewall rule ID matches**.
```

File: products/logs/src/content/instant-logs.md [hunk 4]
```diff
@@ -19,10 +19,119 @@ Instant Logs allows Cloudflare Enterprise customers to access a live stream of t
...
+    "fields": "ClientIP,ClientRequestHost,ClientRequestMethod,ClientRequestURI,EdgeEndTimestamp,EdgeResponseBytes,EdgeResponseStatus,EdgeStartTimestamp,RayID",
+    "sample": 1,
+    "filter": "",
+    "kind": "instant-logs"
+}'
+```
+
+Response:
+
+The response will include a new field called **destination_conf**. The value of this field is your unique WebSocket address that will receive messages from the Edge.
```
Comment (marciocloudflare) [inline products/logs/src/content/instant-logs.md]:
```suggestion
The response will include a new field called **destination_conf**. The value of this field is your unique WebSocket address that will receive messages from Cloudflare's edge.
```

File: products/logs/src/content/instant-logs.md [hunk 5]
```diff
@@ -19,10 +19,119 @@ Instant Logs allows Cloudflare Enterprise customers to access a live stream of t
...
+websocat wss://datalab.cfdata.org/instant-logs/ws/sessions/949f9eb846f06d8f8b7c91b186a349d2
+```
+
+Response:
+
+Once connected to the websocket, you will receive messages of line-delimited JSON.
+
+### Angle Grinder
+
+Now that you have a connection to our websocket and are now receiving logs from the edge, you can start slicing and dicing the logs. A handy tool to use for this is https://github.com/rcoh/angle-grinder. **Angle Grinder** lets you apply filtering, transformations and aggregations on stdin with first class JSON support. For example, to get the number of visitors from each country we can sum the number of events by the `ClientCountry` field.
```
Comment (marciocloudflare) [inline products/logs/src/content/instant-logs.md]:
```suggestion
Now that you have a connection to Cloudflare's websocket and are receiving logs from the edge, you can start slicing and dicing the logs. A handy tool for this is [Angle Grinder](https://github.com/rcoh/angle-grinder). Angle Grinder lets you apply filtering, transformations and aggregations on stdin with first class JSON support. For example, to get the number of visitors from each country you can sum the number of events by the `ClientCountry` field.
```

Comment (marciocloudflare) [reply products/logs/src/content/instant-logs.md]: Not entirely sure if "stin" should be in between backticks or not

File: products/logs/src/content/instant-logs.md [hunk 6]
```diff
@@ -33,13 +142,27 @@ You can download the table of logs that appears in your dash using the **Export*
 
 ## Limits
 
-Instant Logs has two limits set in place:
+Instant Logs has three limits set in place:
 
 * Only one active Instant Logs session per zone.
 * Maximum session time is 60 minutes.
+* If you stop listening to a socket for > 5 mins.
```
Comment (marciocloudflare) [inline products/logs/src/content/instant-logs.md]:
```suggestion
* If you stop listening to a socket for more than five minutes.
```

File: products/logs/src/content/instant-logs.md [hunk 7]
```diff
@@ -33,13 +142,27 @@ You can download the table of logs that appears in your dash using the **Export*
...
+- Get a HTTP 301 when attempting to connect to the WebSocket.
+    -  Make sure you are using the `wss://` protocol when connecting to your WebSocket.
+
+- Get a HTTP 429.
+    - Connection requests are rate limited. Try your request again after waiting a few minutes.
+
+- No data received.
+    -  First, double check if you have a filter defined. If you do, it may be too strict (or incorrect) which ends up dropping all your data.
+    - If you are confident in your filter, check the sample rate you used when creating the session. For example, a sample of 100 means you will receive one log for every 100 requests to your zone.
+    - Finally, make sure the destination is orange clouded. We cannot log your request if it does not go through our edge network.
```
Comment (marciocloudflare) [inline products/logs/src/content/instant-logs.md]:
```suggestion
    - Finally, make sure the destination is proxied through Cloudflare (also known as orange clouded). We cannot log your request if it does not go through our edge network.
```

**Review by marciocloudflare — APPROVED** [review]

**Review by syahril920827 — APPROVED** [review]

File: products/logs/src/content/instant-logs.md [hunk 8]
```diff
@@ -19,10 +19,121 @@ Instant Logs allows Cloudflare Enterprise customers to access a live stream of t
...
+```
+
+Example request using cURL:
+
+```bash
+curl -X POST 'https://api.cloudflare.com/client/v4/zones/${ZONE_ID}/logpush/edge/jobs' \
+-H 'X-Auth-Key: <KEY>' \
+-H 'X-Auth-Email: <EMAIL>' \
+-H 'Content-Type: application/json' \
+- '{
```
Comment (tanushree-sharma) [inline products/logs/src/content/instant-logs.md]: Should be `-d`

File: products/logs/src/content/instant-logs.md [hunk 9]
```diff
@@ -19,10 +19,121 @@ Instant Logs allows Cloudflare Enterprise customers to access a live stream of t
...
+| Greater Than | `"gt"`     |
+| Greater Than or Equal to  | `"geq"`  |
+| Less Than      | `"lt"`   |
+| Less Than or Equal to  | `"leq"`     |
+| Starts with      | `"startsWith"`    |
+| Ends with  | `"endsWith"`            |
+| Contains      | `"contains"`         |
+| Is in  | `"In"`            |
+
+Below we have three examples of requests:
```
Comment (tanushree-sharma) [inline products/logs/src/content/instant-logs.md]: "Below we have three examples of filters:"

File: products/logs/src/content/instant-logs.md [hunk 10]
```diff
@@ -19,10 +19,121 @@ Instant Logs allows Cloudflare Enterprise customers to access a live stream of t
...
+curl -X POST 'https://api.cloudflare.com/client/v4/zones/${ZONE_ID}/logpush/edge/jobs' \
+-H 'X-Auth-Key: <KEY>' \
+-H 'X-Auth-Email: <EMAIL>' \
+-H 'Content-Type: application/json' \
+- '{
+    "fields": "ClientIP,ClientRequestHost,ClientRequestMethod,ClientRequestURI,EdgeEndTimestamp,EdgeResponseBytes,EdgeResponseStatus,EdgeStartTimestamp,RayID",
+    "sample": 1,
+    "filter": "",
+    "kind": "instant-logs"
+}'
```
Comment (tanushree-sharma) [inline products/logs/src/content/instant-logs.md]: Are we staying away from adding `| jq .` on curl calls? It really helps the readability of the API response so I would like to add it in if that's an option

File: products/logs/src/content/instant-logs.md [hunk 11]
```diff
@@ -19,10 +19,121 @@ Instant Logs allows Cloudflare Enterprise customers to access a live stream of t
...
+| Less Than or Equal to  | `"leq"`     |
+| Starts with      | `"startsWith"`    |
+| Ends with  | `"endsWith"`            |
+| Contains      | `"contains"`         |
+| Is in  | `"In"`            |
+
+Below we have three examples of requests:
+
+```bash
+"filter":"{"where":{"and":[{"key":"ClientCountry","operator":"neq","value":"ca"}]}}"
```
Comment (tanushree-sharma) [inline products/logs/src/content/instant-logs.md]: Updating the filters (I have the wrong syntax here) 
`"{\"where\":{\"and\":[{\"key\":\"ClientCountry\",\"operator\":\"neq\",\"value\":\"ca\"}]}}"`

File: products/logs/src/content/instant-logs.md [hunk 12]
```diff
@@ -19,10 +19,121 @@ Instant Logs allows Cloudflare Enterprise customers to access a live stream of t
...
+```
+
+```bash
+"filter":"{"where":{"and":[{"key":"EdgeResponseStatus","operator":"in","value":"200,201"}]}}"
```
Comment (tanushree-sharma) [inline products/logs/src/content/instant-logs.md]: `"{\"where\":{\"and\":[{\"key\":\"EdgeResponseStatus\",\"operator\":\"in\",\"value\":\"200,201\"}]}}"`

File: products/logs/src/content/instant-logs.md [hunk 13]
```diff
@@ -19,10 +19,121 @@ Instant Logs allows Cloudflare Enterprise customers to access a live stream of t
...
+```
+
+```bash
+"filter":"{"where":{"and":[{"key":"ClientRequestPath","operator":"contains","value":"/static"},{"key":"ClientRequestHost","operator":"eq","value":"theburritobot.com"}]}}"
```
Comment (tanushree-sharma) [inline products/logs/src/content/instant-logs.md]: `"{\"where\":{\"and\":[{\"key\":\"ClientRequestPath\",\"operator\":\"contains\",\"value\":\"/static\"}, {\"where\":{\"and\":[{\"key\":\"ClientRequestHost\",\"operator\":\"eq\",\"value\":\"theburritobot.com\"}]}}"`

File: products/logs/src/content/instant-logs.md [hunk 14]
```diff
@@ -19,10 +19,121 @@ Instant Logs allows Cloudflare Enterprise customers to access a live stream of t
...
+
+* **Fields** - List any field available in our [HTTP request dataset](https://developers.cloudflare.com/logs/reference/log-fields/zone/http_requests).
+
+* **Sample** - The sample parameter is the sample rate of the records set by the client: `"sample": 1` is 100% of records `"sample": 10` is 10% and so on.
+
+<Aside type='note'>
+Instant Logs has a maximum data rate supported. For high volume domains, we sample server side as indicated in the `"sampleInterval"` parameter returned in the logs.
+</Aside>
+
+* **Filters** - Use filters to drill down into specific events. Filters consist of three parts: key, operator and value. The keys we support are **ASN**, **Cache status**, **Country**, **Client IP**, **Host**, **HTTP method**, **Path**, **Status code**, **Firewall action matches**, and **Firewall rule ID matches**.
```
Comment (tanushree-sharma) [inline products/logs/src/content/instant-logs.md]: We should have these match the actual field names: `Client ASN, CacheCacheStatus, ClientCountry, ClientIP, ClientRequestHost, ClientRequestPath, EdgeResponseStatus, FirewallMatchesAction, FirewallMatchesRuleIDs`

File: products/logs/src/content/instant-logs.md [hunk 15]
```diff
@@ -19,10 +19,121 @@ Instant Logs allows Cloudflare Enterprise customers to access a live stream of t
...
+        "destination_conf": "wss://datalab.cfdata.org/instant-logs/ws/sessions/949f9eb846f06d8f8b7c91b186a349d2",
+        "kind": "instant-logs"
+    },
+    "success": true
+}
+```
+
+### 2. Connect to WebSocket
+
+Using a CLI utility like Websocat, you can connect to the WebSocket and start immediately receiving logs.
```
Comment (tanushree-sharma) [inline products/logs/src/content/instant-logs.md]: Can we link "Websocat" to their Github repo https://github.com/vi/websocat

File: products/logs/src/content/instant-logs.md [hunk 16]
```diff
@@ -19,10 +19,121 @@ Instant Logs allows Cloudflare Enterprise customers to access a live stream of t
...
+
+### Angle Grinder
+
+Now that you have a connection to Cloudflare's websocket and are receiving logs from the edge, you can start slicing and dicing the logs. A handy tool for this is [Angle Grinder](https://github.com/rcoh/angle-grinder). Angle Grinder lets you apply filtering, transformations and aggregations on stdin with first class JSON support. For example, to get the number of visitors from each country you can sum the number of events by the `ClientCountry` field.
+
+```bash
+websocat wss://datalab.cfdata.org/instant-logs/ws/sessions/949f9eb846f06d8f8b7c91b186a349d2 | agrind '* | json | sum(sampleInterval) by ClientCountry'
+```
+
+Here are a few examples of **ClientCountry** codes:
```
Comment (tanushree-sharma) [inline products/logs/src/content/instant-logs.md]: Maybe replace this with "Response"

File: products/logs/src/content/instant-logs.md [hunk 17]
```diff
@@ -33,13 +144,27 @@ You can download the table of logs that appears in your dash using the **Export*
...
+    -  Make sure you are using the `wss://` protocol when connecting to your WebSocket.
+
+- Get a HTTP 429.
+    - Connection requests are rate limited. Try your request again after waiting a few minutes.
+
+- No data received.
+    -  First, double check if you have a filter defined. If you do, it may be too strict (or incorrect) which ends up dropping all your data.
+    - If you are confident in your filter, check the sample rate you used when creating the session. For example, a sample of 100 means you will receive one log for every 100 requests to your zone.
+    - Finally, make sure the destination is proxied through Cloudflare (also known as orange clouded). We cannot log your request if it does not go through our edge network.
+
```
Comment (tanushree-sharma) [inline products/logs/src/content/instant-logs.md:167]: One more to add here:
- There was an error fetching your data
      -  Make sure you have the correct permissions. In order to use Instant Logs you need Super Administrator, Administrator, Log Share or Log Share Reader permissions.

**Review by tanushree-sharma — CHANGES_REQUESTED** [review]

**Review by soheiokamoto — APPROVED** [review]

**Review by tanushree-sharma — APPROVED** [review]

**Pull request merged**